File: go/sskr/encoding.go
package sskr

import (
	bcrand "github.com/nickel-blockchaincommons/bcrand-go"
	bcshamir "github.com/nickel-blockchaincommons/bcshamir-go"
)

// SSKRGenerate generates SSKR shares for the given spec and master secret.
func SSKRGenerate(spec *Spec, masterSecret *Secret) ([][][]byte, error) {
	rng := bcrand.NewSecureRandomNumberGenerator()
	return SSKRGenerateUsing(spec, masterSecret, rng)
}

// SSKRGenerateUsing generates SSKR shares using a supplied random generator.
func SSKRGenerateUsing(
	spec *Spec,
	masterSecret *Secret,
	randomGenerator bcrand.RandomNumberGenerator,
) ([][][]byte, error) {
	groupsShares, err := generateShares(spec, masterSecret, randomGenerator)
	if err != nil {
		return nil, err
	}

	result := make([][][]byte, len(groupsShares))
	for i, group := range groupsShares {
		serialized := make([][]byte, len(group))
		for j, share := range group {
			serialized[j] = serializeShare(share)
		}
		result[i] = serialized
	}

	return result, nil
}

// SSKRCombine combines SSKR shares into the original secret.
func SSKRCombine(shares [][]byte) (Secret, error) {
	sskrShares := make([]sskrShare, 0, len(shares))
	for _, share := range shares {
		deserialized, err := deserializeShare(share)
		if err != nil {
			return Secret{}, err
		}
		sskrShares = append(sskrShares, deserialized)
	}
	return combineShares(sskrShares)
}

func serializeShare(share sskrShare) []byte {
	// pack the id, group and member data into 5 bytes:
	// 76543210        76543210        76543210
	//         76543210        76543210
	// ----------------====----====----====----
	// identifier: 16
	//                 group-threshold: 4
	//                     group-count: 4
	//                         group-index: 4
	//                             member-threshold: 4
	//                                 reserved (MUST be zero): 4
	//                                     member-index: 4

	result := make([]byte, 0, share.value.Len()+MetadataSizeBytes)
	id := share.identifier
	gt := (share.groupThreshold - 1) & 0xf
	gc := (share.groupCount - 1) & 0xf
	gi := share.groupIndex & 0xf
	mt := (share.memberThreshold - 1) & 0xf
	mi := share.memberIndex & 0xf

	result = append(result,
		byte(id>>8),
		byte(id&0xff),
		byte((gt<<4)|gc),
		byte((gi<<4)|mt),
		byte(mi),
	)
	result = append(result, share.value.bytes()...)

	return result
}

func deserializeShare(source []byte) (sskrShare, error) {
	if len(source) < MetadataSizeBytes {
		return sskrShare{}, ErrShareLengthInvalid
	}

	groupThreshold := int((source[2] >> 4) + 1)
	groupCount := int((source[2] & 0xf) + 1)
	if groupThreshold > groupCount {
		return sskrShare{}, ErrGroupThresholdInvalid
	}

	identifier := (uint16(source[0]) << 8) | uint16(source[1])
	groupIndex := int(source[3] >> 4)
	memberThreshold := int((source[3] & 0xf) + 1)
	reserved := source[4] >> 4
	if reserved != 0 {
		return sskrShare{}, ErrShareReservedBitsInvalid
	}
	memberIndex := int(source[4] & 0xf)

	value, err := NewSecret(source[MetadataSizeBytes:])
	if err != nil {
		return sskrShare{}, err
	}

	return newSSKRShare(
		identifier,
		groupIndex,
		groupThreshold,
		groupCount,
		memberIndex,
		memberThreshold,
		value,
	), nil
}

func generateShares(
	spec *Spec,
	masterSecret *Secret,
	randomGenerator bcrand.RandomNumberGenerator,
) ([][]sskrShare, error) {
	identifierBytes := [2]byte{}
	randomGenerator.FillRandomData(identifierBytes[:])
	identifier := (uint16(identifierBytes[0]) << 8) | uint16(identifierBytes[1])

	groupSecrets, err := bcshamir.SplitSecret(
		spec.GroupThreshold(),
		spec.GroupCount(),
		masterSecret.bytes(),
		randomGenerator,
	)
	if err != nil {
		return nil, wrapShamirError(err)
	}

	groupsShares := make([][]sskrShare, 0, spec.GroupCount())
	for groupIndex, group := range spec.Groups() {
		groupSecret := groupSecrets[groupIndex]
		memberSecretsRaw, err := bcshamir.SplitSecret(
			group.MemberThreshold(),
			group.MemberCount(),
			groupSecret,
			randomGenerator,
		)
		if err != nil {
			return nil, wrapShamirError(err)
		}

		memberShares := make([]sskrShare, 0, len(memberSecretsRaw))
		for memberIndex, memberSecretRaw := range memberSecretsRaw {
			memberSecret, err := NewSecret(memberSecretRaw)
			if err != nil {
				return nil, err
			}
			memberShares = append(memberShares, newSSKRShare(
				identifier,
				groupIndex,
				spec.GroupThreshold(),
				spec.GroupCount(),
				memberIndex,
				group.MemberThreshold(),
				memberSecret,
			))
		}
		groupsShares = append(groupsShares, memberShares)
	}

	return groupsShares, nil
}

type recoveryGroup struct {
	groupIndex      int
	memberThreshold int
	memberIndexes   []int
	memberShares    []Secret
}

func newRecoveryGroup(groupIndex int, memberThreshold int) recoveryGroup {
	return recoveryGroup{
		groupIndex:      groupIndex,
		memberThreshold: memberThreshold,
		memberIndexes:   make([]int, 0, 16),
		memberShares:    make([]Secret, 0, 16),
	}
}

func combineShares(shares []sskrShare) (Secret, error) {
	if len(shares) == 0 {
		return Secret{}, ErrSharesEmpty
	}

	identifier := uint16(0)
	groupThreshold := 0
	groupCount := 0
	secretLen := 0

	groups := make([]recoveryGroup, 0, 16)

	for i, share := range shares {
		if i == 0 {
			identifier = share.identifier
			groupCount = share.groupCount
			groupThreshold = share.groupThreshold
			secretLen = share.value.Len()
		} else {
			if share.identifier != identifier ||
				share.groupThreshold != groupThreshold ||
				share.groupCount != groupCount ||
				share.value.Len() != secretLen {
				return Secret{}, ErrShareSetInvalid
			}
		}

		groupFound := false
		for j := range groups {
			group := &groups[j]
			if share.groupIndex == group.groupIndex {
				groupFound = true
				if share.memberThreshold != group.memberThreshold {
					return Secret{}, ErrMemberThresholdInvalid
				}
				for _, index := range group.memberIndexes {
					if share.memberIndex == index {
						return Secret{}, ErrDuplicateMemberIndex
					}
				}
				if len(group.memberIndexes) < group.memberThreshold {
					group.memberIndexes = append(group.memberIndexes, share.memberIndex)
					group.memberShares = append(group.memberShares, share.value.Clone())
				}
			}
		}

		if !groupFound {
			group := newRecoveryGroup(share.groupIndex, share.memberThreshold)
			group.memberIndexes = append(group.memberIndexes, share.memberIndex)
			group.memberShares = append(group.memberShares, share.value.Clone())
			groups = append(groups, group)
		}
	}

	if len(groups) < groupThreshold {
		return Secret{}, ErrNotEnoughGroups
	}

	masterIndexes := make([]int, 0, 16)
	masterShares := make([][]byte, 0, 16)

	for _, group := range groups {
		if len(group.memberIndexes) < group.memberThreshold {
			continue
		}

		memberShares := make([][]byte, len(group.memberShares))
		for i := range group.memberShares {
			memberShares[i] = group.memberShares[i].bytes()
		}

		groupSecret, err := bcshamir.RecoverSecret(group.memberIndexes, memberShares)
		if err == nil {
			masterIndexes = append(masterIndexes, group.groupIndex)
			masterShares = append(masterShares, groupSecret)
		}

		if len(masterIndexes) == groupThreshold {
			break
		}
	}

	if len(masterIndexes) < groupThreshold {
		return Secret{}, ErrNotEnoughGroups
	}

	masterSecretRaw, err := bcshamir.RecoverSecret(masterIndexes, masterShares)
	if err != nil {
		return Secret{}, wrapShamirError(err)
	}

	masterSecret, err := NewSecret(masterSecretRaw)
	if err != nil {
		return Secret{}, err
	}
	return masterSecret, nil
}

File: typescript/sskr/src/encoding.ts
import type { RandomNumberGenerator } from '@bc/rand';
import { SecureRandomNumberGenerator } from '@bc/rand';
import { ShamirError, splitSecret, recoverSecret } from '@bc/shamir';

import { SskrError } from './error.js';
import { Secret } from './secret.js';
import { SSKRShare } from './share.js';
import type { Spec } from './spec.js';
import { METADATA_SIZE_BYTES } from './constants.js';

/**
 * Generates SSKR shares for the given specification and secret.
 *
 * @param spec - The specification that defines group and member thresholds.
 * @param masterSecret - The secret to split into shares.
 * @returns A nested array of share byte arrays, grouped by SSKR group.
 * @throws {SskrError} If the specification or secret is invalid.
 */
export function sskrGenerate(
    spec: Spec,
    masterSecret: Secret,
): Uint8Array[][] {
    const rng = new SecureRandomNumberGenerator();
    return sskrGenerateUsing(spec, masterSecret, rng);
}

/**
 * Generates SSKR shares using a specific random number generator.
 *
 * @param spec - The specification that defines group and member thresholds.
 * @param masterSecret - The secret to split into shares.
 * @param randomGenerator - The random number generator to use.
 * @returns A nested array of share byte arrays, grouped by SSKR group.
 * @throws {SskrError} If the specification or secret is invalid.
 */
export function sskrGenerateUsing(
    spec: Spec,
    masterSecret: Secret,
    randomGenerator: RandomNumberGenerator,
): Uint8Array[][] {
    const groupsShares = generateShares(spec, masterSecret, randomGenerator);
    return groupsShares.map(group =>
        group.map(share => serializeShare(share)),
    );
}

/**
 * Combines SSKR shares to reconstruct the original secret.
 *
 * @param shares - The SSKR shares to combine. Each share is a byte array.
 * @returns The reconstructed secret.
 * @throws {SskrError} If the shares do not meet the necessary quorum.
 */
export function sskrCombine(shares: Uint8Array[]): Secret {
    const sskrShares: SSKRShare[] = [];
    for (const share of shares) {
        sskrShares.push(deserializeShare(share));
    }
    return combineShares(sskrShares);
}

function serializeShare(share: SSKRShare): Uint8Array {
    // pack the id, group and member data into 5 bytes:
    // 76543210        76543210        76543210
    //         76543210        76543210
    // ----------------====----====----====----
    // identifier: 16
    //                 group-threshold: 4
    //                     group-count: 4
    //                         group-index: 4
    //                             member-threshold: 4
    //                                 reserved (MUST be zero): 4
    //                                     member-index: 4

    const data = share.value.data;
    const result = new Uint8Array(data.length + METADATA_SIZE_BYTES);
    const id = share.identifier;
    const gt = (share.groupThreshold - 1) & 0xf;
    const gc = (share.groupCount - 1) & 0xf;
    const gi = share.groupIndex & 0xf;
    const mt = (share.memberThreshold - 1) & 0xf;
    const mi = share.memberIndex & 0xf;

    result[0] = (id >> 8) & 0xff;
    result[1] = id & 0xff;
    result[2] = (gt << 4) | gc;
    result[3] = (gi << 4) | mt;
    result[4] = mi;
    result.set(data, METADATA_SIZE_BYTES);

    return result;
}

function deserializeShare(source: Uint8Array): SSKRShare {
    if (source.length < METADATA_SIZE_BYTES) {
        throw SskrError.shareLengthInvalid();
    }

    const groupThreshold = ((source[2]! >> 4) + 1);
    const groupCount = ((source[2]! & 0xf) + 1);

    if (groupThreshold > groupCount) {
        throw SskrError.groupThresholdInvalid();
    }

    const identifier = (source[0]! << 8) | source[1]!;
    const groupIndex = (source[3]! >> 4);
    const memberThreshold = ((source[3]! & 0xf) + 1);
    const reserved = source[4]! >> 4;
    if (reserved !== 0) {
        throw SskrError.shareReservedBitsInvalid();
    }
    const memberIndex = (source[4]! & 0xf);
    const value = Secret.create(source.slice(METADATA_SIZE_BYTES));

    return new SSKRShare(
        identifier,
        groupIndex,
        groupThreshold,
        groupCount,
        memberIndex,
        memberThreshold,
        value,
    );
}

function generateShares(
    spec: Spec,
    masterSecret: Secret,
    randomGenerator: RandomNumberGenerator,
): SSKRShare[][] {
    // assign a random identifier
    const idBytes = new Uint8Array(2);
    randomGenerator.fillRandomData(idBytes);
    const identifier = (idBytes[0]! << 8) | idBytes[1]!;

    const groupsShares: SSKRShare[][] = [];

    let groupSecrets: Uint8Array[];
    try {
        groupSecrets = splitSecret(
            spec.groupThreshold,
            spec.groupCount,
            masterSecret.data,
            randomGenerator,
        );
    } catch (e) {
        if (e instanceof ShamirError) {
            throw SskrError.shamirError(e);
        }
        throw e;
    }

    for (let groupIndex = 0; groupIndex < spec.groups.length; groupIndex++) {
        const group = spec.groups[groupIndex]!;
        const groupSecret = groupSecrets[groupIndex]!;

        let memberSecretsRaw: Uint8Array[];
        try {
            memberSecretsRaw = splitSecret(
                group.memberThreshold,
                group.memberCount,
                groupSecret,
                randomGenerator,
            );
        } catch (e) {
            if (e instanceof ShamirError) {
                throw SskrError.shamirError(e);
            }
            throw e;
        }

        const memberSskrShares: SSKRShare[] = memberSecretsRaw.map(
            (memberSecretRaw, memberIndex) => {
                const memberSecret = Secret.create(memberSecretRaw);
                return new SSKRShare(
                    identifier,
                    groupIndex,
                    spec.groupThreshold,
                    spec.groupCount,
                    memberIndex,
                    group.memberThreshold,
                    memberSecret,
                );
            },
        );
        groupsShares.push(memberSskrShares);
    }

    return groupsShares;
}

interface CombineGroup {
    groupIndex: number;
    memberThreshold: number;
    memberIndexes: number[];
    memberShares: Secret[];
}

function combineShares(shares: SSKRShare[]): Secret {
    if (shares.length === 0) {
        throw SskrError.sharesEmpty();
    }

    let identifier = 0;
    let groupThreshold = 0;
    let groupCount = 0;
    let secretLen = 0;

    const groups: CombineGroup[] = [];

    for (let i = 0; i < shares.length; i++) {
        const share = shares[i]!;
        if (i === 0) {
            identifier = share.identifier;
            groupCount = share.groupCount;
            groupThreshold = share.groupThreshold;
            secretLen = share.value.length;
        } else {
            if (
                share.identifier !== identifier ||
                share.groupThreshold !== groupThreshold ||
                share.groupCount !== groupCount ||
                share.value.length !== secretLen
            ) {
                throw SskrError.shareSetInvalid();
            }
        }

        // sort shares into member groups
        let groupFound = false;
        for (const group of groups) {
            if (share.groupIndex === group.groupIndex) {
                groupFound = true;
                if (share.memberThreshold !== group.memberThreshold) {
                    throw SskrError.memberThresholdInvalid();
                }
                for (const existingIndex of group.memberIndexes) {
                    if (share.memberIndex === existingIndex) {
                        throw SskrError.duplicateMemberIndex();
                    }
                }
                if (group.memberIndexes.length < group.memberThreshold) {
                    group.memberIndexes.push(share.memberIndex);
                    group.memberShares.push(share.value);
                }
            }
        }

        if (!groupFound) {
            groups.push({
                groupIndex: share.groupIndex,
                memberThreshold: share.memberThreshold,
                memberIndexes: [share.memberIndex],
                memberShares: [share.value],
            });
        }
    }

    // Check that we have enough groups to recover the master secret
    if (groups.length < groupThreshold) {
        throw SskrError.notEnoughGroups();
    }

    // Recover each group secret, then combine to recover the master secret
    const masterIndexes: number[] = [];
    const masterShares: Uint8Array[] = [];

    for (const group of groups) {
        // Only attempt to recover the group secret if we have enough shares
        if (group.memberIndexes.length < group.memberThreshold) {
            continue;
        }
        // Recover the group secret
        try {
            const groupSecret = recoverSecret(group.memberIndexes, group.memberShares.map(s => s.data));
            masterIndexes.push(group.groupIndex);
            masterShares.push(groupSecret);
        } catch {
            // Skip groups that cannot be recovered
        }
        // Stop if we have enough groups to recover the master secret
        if (masterIndexes.length === groupThreshold) {
            break;
        }
    }

    // If we don't have enough groups to recover the master secret, return an error
    if (masterIndexes.length < groupThreshold) {
        throw SskrError.notEnoughGroups();
    }

    // Recover the master secret
    let masterSecretRaw: Uint8Array;
    try {
        masterSecretRaw = recoverSecret(masterIndexes, masterShares);
    } catch (e) {
        if (e instanceof ShamirError) {
            throw SskrError.shamirError(e);
        }
        throw e;
    }

    return Secret.create(masterSecretRaw);
}
